File: 3_toque_toc/solutions/solution.cpp
#include <iostream>
#include <string>
#include <map>
#include <algorithm>
using namespace std;

string tidy_up(string recipe) {
	string alphabet = "abcdefghijklmnopqrstuvwxyz";
	map<char, int> occurences;
	for(int i = 0; i < 26; i++) {
		occurences.insert(pair<char, int>(alphabet[i], 0));
	}
	for(int i = 0; i < recipe.size(); i++) {
		occurences[recipe[i]] += 1;
	}

	string odd = "";
	string left = "";
	for(int i = 0; i < 26; i++) {
		if(occurences[alphabet[i]] % 2) {
			if(odd.size())
				return "-";
			else
				odd += alphabet[i];
		}
		left.resize(left.size() + occurences[alphabet[i]] / 2, alphabet[i]);
	}

	string right = left;
	reverse(right.begin(), right.end());
	return left + odd + right;
}

int main() {
	int N;
	string recipe;
	cin >> N;
	for(int i=0; i < N; i++) {
		cin >> recipe;
		cout << tidy_up(recipe) << endl;
	}
	return 0;
}

File: 3_toque_toc/solutions/solution.py
def tidy_up(recipe):
	"""Take a word and give a symmetric version with the same letters, or "-".
	"""
	alphabet = "abcdefghijklmnopqrstuvwxyz"
	occurences = {letter: 0 for letter in alphabet}
	for letter in recipe:
		occurences[letter] += 1

	odd = ""
	left = ""
	for letter in alphabet:
		if occurences[letter] % 2:
			if odd:
				return "-" # there can be at most one letter in odd number
			else:
				odd = letter
		left += letter * (occurences[letter] // 2)

	return left + odd + left[::-1]


N = int(input())
recipes = [input() for _ in range(N)]

for recipe in recipes:
	print(tidy_up(recipe))